pkg/state/continuity_check_test.go

```diff
@@ -0,0 +1,179 @@
+
+import (
+	"bufio"
+	"bytes"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/go-logr/logr"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"gopkg.in/yaml.v3"
+
+	mellanoxv1alpha1 "github.com/Mellanox/network-operator/api/v1alpha1"
+	"github.com/Mellanox/network-operator/pkg/render"
+	"github.com/Mellanox/network-operator/pkg/utils"
+)
+
+func extractContainerNamesFromHelmChart(path string) ([]string, error) {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var parsedData map[string]interface{}
+	err = yaml.Unmarshal(uncommentContainerResources(data), &parsedData)
+	if err != nil {
+		return nil, err
+	}
+
+	containerNames := extractContainerNamesFromSubsection(parsedData)
+
+	return containerNames, nil
+}
+
+// uncommentContainerResources iterates through the document and removes '#' comments from containerResources sections
+func uncommentContainerResources(fileData []byte) []byte {
+	var result bytes.Buffer
+	scanner := bufio.NewScanner(bytes.NewReader(fileData))
+	var processNextLines bool
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.TrimSpace(line) == "# containerResources:" {
+			// Remove "# " and set flag to process next lines
+			line = strings.Replace(line, "# ", "", 1)
+			processNextLines = true
+		} else if processNextLines && strings.HasPrefix(strings.TrimSpace(line), "#  ") {
+			// For subsequent lines starting with "#  ", remove "# "
+			line = strings.Replace(line, "# ", "", 1)
+		} else {
+			processNextLines = false
+		}
+
+		result.WriteString(line + "\n")
+	}
+
+	return result.Bytes()
+}
+
+//nolint:gocognit
+func extractContainerNamesFromSubsection(data interface{}) []string {
+	var names []string
+
+	switch v := data.(type) {
+	case []interface{}:
+		for _, item := range v {
+			names = append(names, extractContainerNamesFromSubsection(item)...)
+		}
+	case map[string]interface{}:
+		for key, value := range v {
+			if key == "containerResources" {
+				if resources, ok := value.([]interface{}); ok {
+					for _, resource := range resources {
+						if resMap, ok := resource.(map[string]interface{}); ok {
+							if name, ok := resMap["name"].(string); ok {
+								names = append(names, name)
+							}
+						}
+					}
+				}
+			} else {
+				names = append(names, extractContainerNamesFromSubsection(value)...)
+			}
+		}
+	}
+
+	return names
+}
+
+func extractContainerNamesFromState(manifestsDirectoryPath string, state ManifestRenderer,
+	cr *mellanoxv1alpha1.NicClusterPolicy, logger logr.Logger) ([]string, error) {
+	files, err := utils.GetFilesWithSuffix(manifestsDirectoryPath, render.ManifestFileSuffix...)
+	if err != nil {
+		return nil, err
+	}
+
+	state.SetRenderer(render.NewRenderer(files))
+
+	return ParseContainerNames(state, cr, logger)
+}
+
+var _ = Describe("Continuity check", func() {
+
+	Context("Resource requirements", func() {
+		It("Resource requirements from helm chart should cover all deployable containers", func() {
```

Comment:
if i add a new state, but i forget to update helm and forget to update this test will it fail ?

Reply:
Yes it will. It compares the list of states rendered from manifests with the list in the helm chart. You don’t need to edit this test when adding a new state

Reply:
names from manifests depends on stateMap which is part of the test 
names from chart are extracted from helm chart

not having a state in both (map in test and chart) means that state will be ignored by the test.
so i dont understand how it will work. maybe its the time :)

Reply:
Right 😅 of course you need to add the new state to the test as well. If you don’t update both, the test won’t fail.

Reply:
This guy returns all nicclusterpolicy states: `newNicClusterPolicyStates`

u can use it to get the expected states then if a state is missing from the map, fail

Reply:
Thanks, adapted the test. Not it will fail if you add a new state and not change both the test and helm chart. You still need to edit the NicClusterPolicy object in the test but maybe we can adopt reflection to address that in the future